src/utils/cuda/cuda_utils.cpp

```diff
@@ -0,0 +1,369 @@
+
+public:
+
+    nixlCudaMemCtxImpl() : nixlCudaMemCtx()
+    {    }
+
+    ~nixlCudaMemCtxImpl() override {
+        if (MEM_VMM_DEV == memType) {
+            nixl_status_t status = releaseVmmCudaCtx(_devId);
+            if (NIXL_SUCCESS != status) {
+                NIXL_ERROR << "Failed to release CUDA context";
+            }
+        }
+    }
+
+    nixl_status_t enableAddr(const void *address, uint64_t chkDevId) override;
+
+    nixl_status_t set() override;
+};
+
+#endif
+
+/****************************************
+ * Static nixlCudaPtr functions
+*****************************************/
+
+#ifdef HAVE_CUDA
+
+#define NIXL_CUDA_PTR_CTX_CLASS nixlCudaMemCtxImpl
+#define NIXL_CUDA_PTR_CTX_VRAM_SUPPORT true
+
+#else
+
+#define NIXL_CUDA_PTR_CTX_CLASS nixlCudaMemCtx
+#define NIXL_CUDA_PTR_CTX_VRAM_SUPPORT false
+
+#endif
+
+bool nixlCudaMemCtx::vramIsSupported()
+{
+    return NIXL_CUDA_PTR_CTX_VRAM_SUPPORT;
+}
+
+std::unique_ptr<nixlCudaMemCtx>
+nixlCudaMemCtx::nixlCudaMemCtxInit()
+{
+    // Environment fixup
+    if (getenv("NIXL_DISABLE_CUDA_ADDR_WA")) {
+        // If the workarounf is disabled - return the dummy class
+        NIXL_INFO << "WARNING: disabling CUDA address workaround";
+        return std::make_unique<nixlCudaMemCtx>();
+    } else {
+        return std::make_unique<NIXL_CUDA_PTR_CTX_CLASS>();
+    }
+}
+
+
+/****************************************
+ * CUDA nixlCudaPtr class implementation
+*****************************************/
+
+#ifdef HAVE_CUDA
+
+nixl_status_t
+nixlCudaMemCtxImpl::queryVmm(const void *address, memory_t &type, uint64_t &id)
+{
+    nixl_status_t ret = NIXL_ERR_NOT_FOUND;
+
+#if HAVE_CUMEMRETAINALLOCATIONHANDLE
+    CUmemAllocationProp prop = {};
+    CUmemGenericAllocationHandle alloc_handle;
+    CUresult result;
+
+    /* Check if memory is allocated using VMM API and see if host memory needs
+     * to be treated as pinned device memory */
+    result = cuMemRetainAllocationHandle(&alloc_handle, (void*)address);
+    if (result != CUDA_SUCCESS) {
+        NIXL_DEBUG << "cuMemRetainAllocationHandle() failed. result = "
+                   << result;
+        return NIXL_ERR_NOT_FOUND;
+    }
+    // TODO: set the call to cuMemRelease when leaving the scope to avoid GOTO
+
+    result = cuMemGetAllocationPropertiesFromHandle(&prop, alloc_handle);
+    if (result != CUDA_SUCCESS) {
+        NIXL_DEBUG << "cuMemGetAllocationPropertiesFromHandle() failed. result = "
+                   << result;
+        ret = NIXL_ERR_UNKNOWN;
+        goto err;
+    }
+
+    id = (CUdevice)prop.location.id;
+    switch (prop.location.type) {
+    case CU_MEM_LOCATION_TYPE_DEVICE:
+        type = MEM_VMM_DEV;
+        ret = NIXL_SUCCESS;
+        break;
+#if HAVE_DECL_CU_MEM_LOCATION_TYPE_HOST
+    case CU_MEM_LOCATION_TYPE_HOST:
+    case CU_MEM_LOCATION_TYPE_HOST_NUMA:
+    case CU_MEM_LOCATION_TYPE_HOST_NUMA_CURRENT:
+        // Currently not supported
+        //type = MEM_VMM_HOST;
+#endif
```

Comment:
if this is not supported, do we really need to define HAVE_DECL_CU_MEM_LOCATION_TYPE_HOST and add the meson build code for it, and also introduce meson 1.3.0 requirement?